pkg/fqdn/dnsproxy/udp.go

```diff
@@ -0,0 +1,250 @@
+}
+
+type sessionUDPFactory struct{}
+
+var ciliumSessionUDPFactory *sessionUDPFactory
+
+var rawconn4 *net.IPConn // raw socket for sending IPv4
+var rawconn6 *net.IPConn // raw socket for sending IPv6
+
+// Set the socket options needed for tranparent proxying for the listening socket
+// IP(V6)_TRANSPARENT allows socket to receive packets with any destination address/port
+// IP(V6)_RECVORIGDSTADDR tells the kernel to pass the original destination address/port on recvmsg
+// The socket may be receiving both IPv4 and IPv6 data, so set both options, if possible.
+func transparentSetsockopt(fd int) error {
+	err6 := unix.SetsockoptInt(fd, unix.SOL_IPV6, unix.IPV6_TRANSPARENT, 1)
+	if err6 == nil {
+		err6 = unix.SetsockoptInt(fd, unix.SOL_IPV6, unix.IPV6_RECVORIGDSTADDR, 1)
+	}
+	err4 := unix.SetsockoptInt(fd, unix.SOL_IP, unix.IP_TRANSPARENT, 1)
+	if err4 == nil {
+		err4 = unix.SetsockoptInt(fd, unix.SOL_IP, unix.IP_RECVORIGDSTADDR, 1)
+	}
+	// Only error out if both IPv4 and IPv6 fail
+	if err4 != nil && err6 != nil {
+		return err4
+	}
+	return nil
+}
+
+func listenConfig(mark int) *net.ListenConfig {
+	return &net.ListenConfig{
+		Control: func(network, address string, c syscall.RawConn) error {
+			var opErr error
+			err := c.Control(func(fd uintptr) {
+				opErr = transparentSetsockopt(int(fd))
+				if opErr == nil && mark != 0 {
+					opErr = syscall.SetsockoptInt(int(fd), syscall.SOL_SOCKET, syscall.SO_MARK, mark)
+				}
+				if opErr == nil {
+					opErr = syscall.SetsockoptInt(int(fd), syscall.SOL_SOCKET, syscall.SO_REUSEADDR, 1)
+				}
+			})
+			if err != nil {
+				return err
+			}
+
+			return opErr
+		}}
+}
+
+func bindUDP(addr string) *net.IPConn {
+	// Mark outgoing packets as proxy egress return traffic (0x0b00)
+	conn, err := listenConfig(0xb00).ListenPacket(context.Background(), "ip:udp", addr)
+	if err != nil {
+		log.Errorf("bindUDP failed for address %s: %s", addr, err)
+		return nil
+	}
+	return conn.(*net.IPConn)
+}
+
+func (f *sessionUDPFactory) CreateSessionUDP(msgSize int) dns.SessionUDP {
+	return &sessionUDP{
+		m:   make([]byte, msgSize),
+		oob: make([]byte, udpOOBSize),
+	}
+}
+
+var udpOnce sync.Once
+
+func (s *sessionUDPFactory) InitConn(conn *net.UDPConn) error {
+	// Set up the raw socket for sending responses.
+	// - Must use a raw UDP socket for sending responses so that we can send
+	//   from a specific port without binding to it.
+	// - The raw UDP socket must be bound to a specific IP address to prevent
+	//   it receiving ALL UDP packets on the host.
+	// - We use oob data to override the source IP address when sending
+	// - Must use separate sockets for IPv4/IPv6, as sending to a v6-mapped
+	//   v4 address from a socket bound to "::1" does not work due to kernel
+	//   checking that a route exists from the source address before
+	//   the source address is replaced with the (transparently) changed one
+	udpOnce.Do(func() {
+		rawconn4 = bindUDP("127.0.0.1") // raw socket for sending IPv4
```

Comment:
same thing that joe said, you can initialize this directly in this file

Reply:
Done.